### PR TITLE
Handle the case where an initializer has a composition of types

### PR DESCRIPTION
**Pull request by fdiaz**
Work on https://github.com/fdiaz/SwiftInspector/issues/20

This fixes an edge case of when an initializer argument contains a composition of elements.

### CODE REVIEW

File: Sources/SwiftInspectorCommands/InitializerCommand.swift [hunk 1]
```diff
@@ -71,9 +71,9 @@ final class InitializerCommand: ParsableCommand {
 
   private func outputString(from statement: InitializerStatement) -> String {
     if typeOnly {
-      return statement.parameters.map { $0.typeName }.joined(separator: " ")
+      return statement.parameters.map { $0.typeNames.joined(separator: " ") }.joined(separator: " ")
```
Comment (bachand) [inline Sources/SwiftInspectorCommands/InitializerCommand.swift:74]: I'm a bit confused by how you parse the output now. It seems like it would be hard to parse `foo,MyProtocol YourProtocol anotherParam:String`.

Should it be `foo,MyProtocol,YourProtocol anotherParam:String` instead?

Comment (bachand) [inline Sources/SwiftInspectorCommands/InitializerCommand.swift:74]: Also while we are thinking about edge cases like this, what happens if the type is a tuple? Or a closure?

File: Sources/SwiftInspectorCore/InitializerAnalyzer.swift [hunk 2]
```diff
@@ -180,3 +203,17 @@ private final class InitializerSyntaxReader: SyntaxVisitor {
   let shouldVisitIdentifier: (String) -> Bool
   let onNodeVisit: (InitializerDeclSyntax) -> Void
 }
+
+private final class FunctionParameterReader: SyntaxVisitor {
```
Comment (bachand) [inline Sources/SwiftInspectorCore/InitializerAnalyzer.swift:207]: I like the second syntax visitor approach

File: Sources/SwiftInspectorCommands/InitializerCommand.swift (hunk 1 above)
Comment (bachand) [reply Sources/SwiftInspectorCommands/InitializerCommand.swift:74]: In the case where you exclude parameter names, how do you know which types belong to switch initializer parameter? How do I parse `MyProtocol YouProtocol String`?

Comment (fdiaz) [reply Sources/SwiftInspectorCommands/InitializerCommand.swift:74]: @bachand this is for the type only case. So this would output:

`MyProtocol YourProtocol String` in your use case.

[Below](https://github.com/fdiaz/SwiftInspector/pull/21/files#diff-f8fee815716808d2efcdd15e9654898aR76) is the update for multiple types when outputting the argument name. It'd look like:

`foo,MyProtocol,YourProtocol anotherParam,String`

Comment (fdiaz) [reply Sources/SwiftInspectorCommands/InitializerCommand.swift:74]: Good call! Let me add more tests and try those use cases.

File: Sources/SwiftInspectorCore/InitializerAnalyzer.swift (hunk 2 above)
Comment (fdiaz) [reply Sources/SwiftInspectorCore/InitializerAnalyzer.swift:207]: Thanks! Yeah I think it makes it a lot cleaner to follow the code.

File: Sources/SwiftInspectorCommands/InitializerCommand.swift (hunk 1 above)
Comment (fdiaz) [reply Sources/SwiftInspectorCommands/InitializerCommand.swift:74]: For a tuple we're handling this correctly. For a closure we're just returning the types on the closure itself but it's somewhat weird of a syntax since it'll just list the types.

I created this https://github.com/fdiaz/SwiftInspector/issues/23 to track this since I think this might be a bigger change.

Comment (fdiaz) [reply Sources/SwiftInspectorCommands/InitializerCommand.swift:74]: typle commit https://github.com/fdiaz/SwiftInspector/pull/22/commits/a12e7637a8120e5c9ee2ac0a6761767ae0ebca64

Comment (bachand) [reply Sources/SwiftInspectorCommands/InitializerCommand.swift:74]: @fdiaz I'm curious why you chose to output `MyProtocol YourProtocol String` in the type only case vs. `MyProtocol,YourProtocol String`. In any case, I'm fine deferring this until we address https://github.com/fdiaz/SwiftInspector/issues/23

Comment (fdiaz) [reply Sources/SwiftInspectorCommands/InitializerCommand.swift:74]: I think its easier to parse when we do that, and if we're saying we only care about the types I don't see a value in adding context for each argument.

What I'm guessing this can be used to is to know what types a certain class takes in case there are types we want to forbid or warn about.

Since we have the `--disable-type-only` flag, if we do care about the parameters and how they're grouped together we can always get that information from there instead.

Comment (bachand) [reply Sources/SwiftInspectorCommands/InitializerCommand.swift:74]: > What I'm guessing this can be used to is to know what types a certain class takes in case there are types we want to forbid or warn about.

Makes sense!
